### DISCUSSION TITLE
BadUSB/BT enter at the end?

### DISCUSSION THREAD

**Discussion by eried** [akopachov/flipper-zero_authenticator#94] [Ideas]
Hi! have you considered sending an enter at the end? yes/no? reasons for no?

(im loving the BT keyboard functionality, but I wonder what would be the drawback of having enter pressed... maybe double tap up could send BT with enter)

**Comment by eried** [comment]
(also offtopic, your ko-fi link is not working, I had 2 people trying and gived paypal error)

**Reply by akopachov** [reply]
Hey, thanks for reporting this. I actually never tried donated myself, so I didn't even knew it is broken, lol
I turned off that paypal option and enabled "Pay with card" and "Google Pay \ Apple Pay" option. Should be much more convenient.

**Comment by akopachov** [comment]
> Hi! have you considered sending an enter at the end? yes/no? reasons for no?
> 
> (im loving the BT keyboard functionality, but I wonder what would be the drawback of having enter pressed... maybe double tap up could send BT with enter)

Initial implementation of BadUSB did that "enter key" press at the end. But then once more people start using it I received multiple reports saying that not all the website requires user to confirm code by "enter" key, instead they are just waiting until user enters all the required digits.

So I dropped that "enter" key at the end to make sure that it will not produce an issues  on such a websites. After that nobody complained and nobody wanted to return that "enter" key back, so I just let it stay as is.

**Reply by akopachov** [reply]
While typing this answer, I start wondering if it might be a good idea to add sort of "configuration" to let user decide whether to press "enter" key at the end or not.

**Reply by eried** [reply]
maybe, or per key config, it could be just that the name ends with a special char like * (even if is not displayed). But yes some websites dont need enter, but does it affect? like there is a spinner animation after typing the last char no? (it is like old people double clicking a hyperlink)

**Reply by akopachov** [reply]
Okay, you convinced me 😄 Per-key config should make it flexible enough and make happy everybody. Logged [TODO item](https://github.com/users/akopachov/projects/1/views/1?pane=issue&itemId=23808692) for this idea.

**Reply by eried** [reply]
I kinda think the bluetooth icon looks odd :P not sure why. Maybe is because it is alone in the middle. It could have some leyend about the functionality
![image](https://user-images.githubusercontent.com/1091420/227520775-0c9bb0da-ff4d-426b-8f2c-51fc6d24e17d.png)

and hold down for usb

**Reply by eried** [reply]
oh actually something like this would look better

![image](https://user-images.githubusercontent.com/1091420/227521620-4e39bfba-293b-4e27-8f08-4b94363cb1d7.png)

it even first on the 10 pixel height :))

**Reply by akopachov** [reply]
> I kinda think the bluetooth icon looks odd :P

That is why there is FAP without this icon 🤣 

I like the idea though. Will play with it more

**Reply by akopachov** [reply]
How does it look now?

![Screenshot-20230327-151814](https://user-images.githubusercontent.com/1217426/227948730-eae13705-5190-4ee7-a599-8b1b0f56daba.png)

**Reply by eried** [reply]
I think looks majestic 😇 difficult to convey the usb too in little space, the cactus haha

**Reply by akopachov** [reply]
ha-ha! Yeah, indeed a cactus, lol

**Reply by eried** [reply]
![image](https://user-images.githubusercontent.com/1091420/227960331-b4172c5d-e422-42fb-bb12-ec32a0162cef.png)

I dont know, it is mostly opinion in so little pixels

**Reply by eried** [reply]
depends how it looks in the screen too

**Reply by akopachov** [reply]
Okay, so it is either cactus or fat mouse, lol

![Screenshot-20230327-160202](https://user-images.githubusercontent.com/1217426/227962306-f526876a-fb9a-4bb2-b942-553bdefbb1d4.png)

**Reply by eried** [reply]
lol cactus is good too, or maybe just 1 keycap from keyboard instead

**Reply by eried** [reply]
![image](https://user-images.githubusercontent.com/1091420/227968084-10e56217-d661-4fa4-88d8-bd995f088ce5.png)

or usb-c

**Reply by eried** [reply]
![image](https://user-images.githubusercontent.com/1091420/227968262-35a76e07-89c9-40be-971b-80197c05f4bd.png)

**Reply by akopachov** [reply]
One more idea

![Screenshot-20230327-162406](https://user-images.githubusercontent.com/1217426/227968937-607d3519-8e87-4493-8547-ff674a2db08a.png)

**Reply by eried** [reply]
yeah that flowerpot vase might work

**Reply by akopachov** [reply]
Let me create a poll in discord, lol

**Reply by eried** [reply]
your discord link does not work btw, it should be invitation not just the link of the channel

**Reply by akopachov** [reply]
Ah, thanks for catching this. Will fix soon

**Reply by akopachov** [reply]
Updated discord link in README.

[Here is](https://discord.com/channels/937479784148115456/1020428040636993536/1090165140290277456) a direct link to poll message

**Comment by eried** [comment]
This works great in latest! extra arguments are awesome! I could imagine tab-> enter for other scenarios

**Reply by akopachov** [reply]
In [v1.8.1](https://github.com/akopachov/flipper-zero_authenticator/releases/tag/v1.8.1) there is an ability to turn "Enter tab at the end" feature. It is even possible to turn on both features, <enter> and <tab> for the same token 😄 that case it will type <enter> then <tab>.

**Reply by eried** [reply]
awesome! the features per token is brilliant. Could you add some "slow start" or "slow between keys" modifier? I noticed for some apps sometimes it looses the first keystroke in BT (paypal on android) for example. Im trying to use this auth the whole day

**Reply by akopachov** [reply]
"Slow start" is something what user can handle himself much better, I believe, as Auth app doesn't know if another app is ready to accept keystores or not.

"Slower type" - this is actually good idea! Logged #106

**Reply by akopachov** [reply]
Added in [v1.8.2](https://github.com/akopachov/flipper-zero_authenticator/releases/tag/v1.8.2)

**Reply by eried** [reply]
Yes good. By slow start I mean the moment the bluetooth device is paired could have a custom delay, maybe in settings. I only experience this on paypal on my phone tho, is like the keyboard is ready but then the first keystroke is omitted. Maybe is something else? Usb works flawless even in the same app

**Reply by eried** [reply]
Btw in trying to use this as the sole auth for everything. And carrying my f0 everywhere

**Reply by akopachov** [reply]
> Yes good. By slow start I mean the moment the bluetooth device is paired could have a custom delay, maybe in settings. I only experience this on paypal on my phone tho, is like the keyboard is ready but then the first keystroke is omitted. Maybe is something else? Usb works flawless even in the same app

Well, BT become available for connecting\pairing as soon as you launch application. When you trigger "BT automation" it just starts sending commands to connected device. So I'm not really sure where to put a "start delay" here, given that as a user you may wait as much as needed before triggering "BT automation". Am I missing something?

Will try to investigate more though.



> Btw in trying to use this as the sole auth for everything. And carrying my f0 everywhere

This gives me priceless feedback on real usecases. Thank you .

**Reply by eried** [reply]
After the device connected? Before sending the first keystroke? I think is 500ms now no?

**Reply by akopachov** [reply]
For USB - yes, there is this 500ms delay, because of the nature how USB automation is working:
1. Turn on USB Keyboard mode
2. Wait 500ms to let host OS recognize it
3. Automagic
4. Turn off USB Keyboard mode

BT automation acts slightly different. It doesn't turn on\off BT modes, because it is not necessary (for USB mode it is critical because while USB is in keyboard mode it is impossible to connect to flipper to use CLI). It turns on BT in BT keyboard mode at the start of the app, and turns it off at the end of the app, so there is no need to wait that 500ms.

That is why I was kinda confused 😄.

Do you still experience that issue even with "type slow" feature?

**Reply by eried** [reply]
oh, ok... error is not easy to trigger tho, but I will continue testing
